// flipper-plugin/lib/ui/DataList.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataList = void 0;
const react_1 = __importStar(require("react"));
const DataFormatter_1 = require("./DataFormatter");
const Layout_1 = require("./Layout");
const antd_1 = require("antd");
const DataTable_1 = require("./data-table/DataTable");
const icons_1 = require("@ant-design/icons");
const theme_1 = require("./theme");
const styled_1 = __importDefault(require("@emotion/styled"));
const useAssertStableRef_1 = require("../utils/useAssertStableRef");
const useMakeStableCallback_1 = require("../utils/useMakeStableCallback");
const { Text } = antd_1.Typography;
exports.DataList = Object.assign(function ({ onSelect: baseOnSelect, selection, className, style, items, onRenderItem, enableArrow, idAttribute, titleAttribute, descriptionAttribute, ...tableProps }) {
    var _a;
    const tableManagerRef = (_a = tableProps.tableManagerRef) !== null && _a !== void 0 ? _a : react_1.createRef();
    useAssertStableRef_1.useAssertStableRef(onRenderItem, 'onRenderItem');
    useAssertStableRef_1.useAssertStableRef(enableArrow, 'enableArrow');
    const onSelect = useMakeStableCallback_1.useMakeStableCallback(baseOnSelect);
    const handleSelect = react_1.useCallback((item) => {
        if (!item) {
            onSelect === null || onSelect === void 0 ? void 0 : onSelect(undefined, undefined);
        }
        else {
            const id = '' + item[idAttribute];
            if (id == null) {
                throw new Error(`No valid identifier for attribute ${idAttribute}`);
            }
            onSelect === null || onSelect === void 0 ? void 0 : onSelect(id, item);
        }
    }, [onSelect, idAttribute]);
    react_1.useEffect(() => {
        var _a, _b;
        if (selection) {
            (_a = tableManagerRef.current) === null || _a === void 0 ? void 0 : _a.selectItemById(selection);
        }
        else {
            (_b = tableManagerRef.current) === null || _b === void 0 ? void 0 : _b.clearSelection();
        }
    }, [selection, tableManagerRef]);
    const dataListColumns = react_1.useMemo(() => [
        {
            key: idAttribute,
            wrap: true,
            onRender(item, selected, index) {
                return onRenderItem ? (onRenderItem(item, selected, index)) : (react_1.default.createElement(exports.DataList.Item, { title: item[titleAttribute], description: item[descriptionAttribute], enableArrow: enableArrow }));
            },
        },
    ], [
        onRenderItem,
        enableArrow,
        titleAttribute,
        descriptionAttribute,
        idAttribute,
    ]);
    return (react_1.default.createElement(Layout_1.Layout.Container, { style: style, className: className, grow: true },
        react_1.default.createElement(DataTable_1.DataTable, { ...tableProps, tableManagerRef: tableManagerRef, records: Array.isArray(items) ? items : undefined, dataSource: Array.isArray(items) ? undefined : items, recordsKey: idAttribute, columns: dataListColumns, onSelect: handleSelect })));
}, {
    Item: react_1.memo(({ title, description, enableArrow }) => {
        return (react_1.default.createElement(Layout_1.Layout.Horizontal, { center: true, grow: true, shrink: true, padv: true },
            react_1.default.createElement(Layout_1.Layout.Container, { grow: true, shrink: true },
                typeof title === 'string' ? (react_1.default.createElement(Text, { strong: true, ellipsis: true }, DataFormatter_1.DataFormatter.format(title))) : (title),
                typeof description === 'string' ? (react_1.default.createElement(Text, { type: "secondary", ellipsis: true }, DataFormatter_1.DataFormatter.format(description))) : (description)),
            enableArrow && (react_1.default.createElement(ArrowWrapper, null,
                react_1.default.createElement(icons_1.RightOutlined, null)))));
    }),
});
exports.DataList.defaultProps = {
    type: 'default',
    scrollable: true,
    enableSearchbar: false,
    enableColumnHeaders: false,
    enableArrow: true,
    enableContextMenu: false,
    enableMultiSelect: false,
    idAttribute: 'id',
    titleAttribute: 'title',
    descriptionAttribute: 'description',
};
exports.DataList.Item.defaultProps = {
    enableArrow: true,
};
const ArrowWrapper = styled_1.default.div({
    flex: 0,
    paddingLeft: theme_1.theme.space.small,
    '.anticon': {
        lineHeight: '14px',
    },
});
