// flipper-plugin/lib/ui/DataFormatter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TruncateHelper = exports.DataFormatter = void 0;
const icons_1 = require("@ant-design/icons");
const antd_1 = require("antd");
const lodash_1 = require("lodash");
const react_1 = __importStar(require("react"));
const FlipperLib_1 = require("../plugin/FlipperLib");
const safeStringify_1 = require("../utils/safeStringify");
const urlRegex_1 = require("../utils/urlRegex");
const index_1 = require("../data-source/index");
const theme_1 = require("./theme");
exports.DataFormatter = {
    defaultFormatter(value) {
        if (react_1.isValidElement(value)) {
            return value;
        }
        switch (typeof value) {
            case 'boolean':
                return value ? 'true' : 'false';
            case 'number':
                return '' + value;
            case 'undefined':
                return '';
            case 'string':
                return value;
            case 'object': {
                if (value === null)
                    return '';
                if (value instanceof Date) {
                    return (value.toTimeString().split(' ')[0] +
                        '.' +
                        lodash_1.pad('' + value.getMilliseconds(), 3, '0'));
                }
                if (value instanceof Map) {
                    return safeStringify_1.safeStringify(Array.from(value.entries()));
                }
                if (value instanceof Set) {
                    return safeStringify_1.safeStringify(Array.from(value.values()));
                }
                return safeStringify_1.safeStringify(value);
            }
            default:
                return '<unrenderable value>';
        }
    },
    truncate(maxLength) {
        return (value) => {
            if (typeof value === 'string' && value.length > maxLength) {
                return react_1.default.createElement(TruncateHelper, { value: value, maxLength: maxLength });
            }
            return value;
        };
    },
    linkify(value) {
        if (typeof value === 'string' && urlRegex_1.urlRegex.test(value)) {
            return react_1.createElement(react_1.Fragment, undefined, ...value.split(urlRegex_1.urlRegex).map((part, index) => index % 2 === 1 ? (react_1.default.createElement(antd_1.Typography.Link, { href: part }, part)) : (part)));
        }
        return value;
    },
    prettyPrintJson(value) {
        if (react_1.isValidElement(value)) {
            return value;
        }
        if (typeof value === 'string' && value.length >= 2) {
            const last = value.length - 1;
            if ((value[0] === '{' && value[last] === '}') ||
                (value[0] === '[' && value[last] === ']')) {
                try {
                    value = JSON.parse(value);
                }
                catch (e) {
                }
            }
        }
        if (typeof value === 'object' &&
            value !== null &&
            (Array.isArray(value) || lodash_1.isPlainObject(value))) {
            try {
                return JSON.stringify(value, null, 2);
            }
            catch (e) {
            }
        }
        return value;
    },
    format(value, formatters) {
        let res = value;
        if (Array.isArray(formatters)) {
            for (const formatter of formatters) {
                res = formatter(res);
            }
        }
        else if (formatters) {
            res = formatters(res);
        }
        return exports.DataFormatter.defaultFormatter(res);
    },
};
function TruncateHelper({ value, maxLength, }) {
    const [collapsed, setCollapsed] = react_1.useState(true);
    const redrawRow = index_1.useTableRedraw();
    return (react_1.default.createElement(react_1.default.Fragment, null,
        collapsed ? value.substr(0, maxLength) : value,
        react_1.default.createElement(antd_1.Button, { onClick: () => {
                setCollapsed((c) => !c);
                redrawRow === null || redrawRow === void 0 ? void 0 : redrawRow();
            }, size: "small", type: "text", style: truncateButtonStyle, icon: collapsed ? react_1.default.createElement(icons_1.CaretRightOutlined, null) : react_1.default.createElement(icons_1.CaretUpOutlined, null) }, collapsed ? `and ${value.length - maxLength} more` : 'collapse'),
        react_1.default.createElement(antd_1.Button, { icon: react_1.default.createElement(icons_1.CopyOutlined, null), onClick: () => {
                var _a;
                (_a = FlipperLib_1.tryGetFlipperLibImplementation()) === null || _a === void 0 ? void 0 : _a.writeTextToClipboard(value);
            }, size: "small", type: "text", style: truncateButtonStyle }, "copy")));
}
exports.TruncateHelper = TruncateHelper;
const truncateButtonStyle = {
    color: theme_1.theme.primaryColor,
    marginLeft: 4,
};
