// flipper-plugin/lib/ui/Highlight.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useHighlighter = exports.HighlightProvider = exports.HighlightContext = void 0;
const styled_1 = __importDefault(require("@emotion/styled"));
const react_1 = __importStar(require("react"));
const lodash_1 = require("lodash");
const Highlighted = styled_1.default.span({
    backgroundColor: '#fcd872',
});
function createHighlightManager(initialText = '') {
    const callbacks = new Set();
    let matches = 0;
    let currentFilter = initialText;
    const Highlight = react_1.memo(({ text }) => {
        const [, setUpdate] = react_1.useState(0);
        const elem = react_1.useRef(null);
        react_1.useEffect(() => {
            function onChange(prevHighlight, newHighlight) {
                var _a, _b, _c, _d;
                const prevIndex = text.toLowerCase().indexOf(prevHighlight);
                const newIndex = text.toLowerCase().indexOf(newHighlight);
                if (prevIndex !== newIndex || newIndex !== -1) {
                    if (newIndex !== -1) {
                        if (++matches === 1) {
                            (_d = (_c = (_b = (_a = elem.current) === null || _a === void 0 ? void 0 : _a.parentElement) === null || _b === void 0 ? void 0 : _b.parentElement) === null || _c === void 0 ? void 0 : _c.scrollIntoView) === null || _d === void 0 ? void 0 : _d.call(_c);
                        }
                    }
                    setUpdate((s) => s + 1);
                }
            }
            callbacks.add(onChange);
            return () => {
                callbacks.delete(onChange);
            };
        }, [text]);
        const index = text.toLowerCase().indexOf(currentFilter);
        return (react_1.default.createElement("span", { ref: elem }, index === -1 ? (text) : (react_1.default.createElement(react_1.default.Fragment, null,
            text.substr(0, index),
            react_1.default.createElement(Highlighted, null, text.substr(index, currentFilter.length)),
            text.substr(index + currentFilter.length)))));
    });
    return {
        setFilter: lodash_1.debounce((text = '') => {
            if (currentFilter !== text) {
                matches = 0;
                const base = currentFilter;
                currentFilter = text.toLowerCase();
                callbacks.forEach((cb) => cb(base, currentFilter));
            }
        }, 100),
        render(text) {
            return react_1.default.createElement(Highlight, { text: text });
        },
    };
}
exports.HighlightContext = react_1.createContext({
    setFilter(_text) {
        throw new Error('Cannot set the filter of a stub highlight manager');
    },
    render(text) {
        return text;
    },
});
function HighlightProvider({ text, children, }) {
    const [highlightManager] = react_1.useState(() => createHighlightManager(text));
    react_1.useEffect(() => {
        highlightManager.setFilter(text);
    }, [text, highlightManager]);
    return (react_1.default.createElement(exports.HighlightContext.Provider, { value: highlightManager }, children));
}
exports.HighlightProvider = HighlightProvider;
function useHighlighter() {
    return react_1.useContext(exports.HighlightContext);
}
exports.useHighlighter = useHighlighter;
