// flipper-plugin/lib/ui/MasterDetail.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MasterDetail = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const DataInspector_1 = require("./data-inspector/DataInspector");
const DataTable_1 = require("./data-table/DataTable");
const DetailSidebar_1 = require("./DetailSidebar");
const Layout_1 = require("./Layout");
const Panel_1 = require("./Panel");
const icons_1 = require("@ant-design/icons");
const antd_1 = require("antd");
const PluginContext_1 = require("../plugin/PluginContext");
const atom_1 = require("../state/atom");
const useAssertStableRef_1 = require("../utils/useAssertStableRef");
function MasterDetail({ dataSource, records, sidebarComponent, sidebarPosition, sidebarSize, onSelect, extraActions, enableMenuEntries, enableClear, isPaused, selection, onClear, ...tableProps }) {
    var _a;
    useAssertStableRef_1.useAssertStableRef(isPaused, 'isPaused');
    useAssertStableRef_1.useAssertStableRef(selection, 'selection');
    const pluginInstance = PluginContext_1.usePluginInstance();
    const { client } = pluginInstance;
    const connected = atom_1.useValue(pluginInstance.client.connected);
    const selectionAtom = selection !== null && selection !== void 0 ? selection : react_1.useState(() => atom_1.createState(undefined))[0];
    const selectedRecord = atom_1.useValue(selectionAtom);
    const tableManagerRef = (_a = tableProps.tableManagerRef) !== null && _a !== void 0 ? _a : react_1.createRef();
    const pausedState = atom_1.useValue(isPaused, false);
    const sidebar = sidebarPosition !== 'none' && selectedRecord && sidebarComponent
        ? react_1.createElement(sidebarComponent, {
            record: selectedRecord,
        })
        : null;
    const handleSelect = react_1.useCallback((record, records) => {
        selectionAtom.set(record);
        onSelect === null || onSelect === void 0 ? void 0 : onSelect(record, records);
    }, [selectionAtom, onSelect]);
    const handleTogglePause = react_1.useCallback(() => {
        isPaused === null || isPaused === void 0 ? void 0 : isPaused.set(!(isPaused === null || isPaused === void 0 ? void 0 : isPaused.get()));
    }, [isPaused]);
    const handleClear = react_1.useCallback(() => {
        handleSelect(undefined, []);
        if (dataSource) {
            dataSource.clear();
            onClear === null || onClear === void 0 ? void 0 : onClear();
        }
        else {
            if (!onClear) {
                throw new Error("onClear must be set when using 'enableClear' and 'records'");
            }
            onClear();
        }
    }, [dataSource, onClear, handleSelect]);
    const handleCreatePaste = react_1.useCallback(() => {
        var _a;
        const selection = (_a = tableManagerRef.current) === null || _a === void 0 ? void 0 : _a.getSelectedItems();
        switch (selection === null || selection === void 0 ? void 0 : selection.length) {
            case undefined:
            case 0:
                return;
            case 1:
                client.createPaste(JSON.stringify(selection[0], null, 2));
                break;
            default:
                client.createPaste(JSON.stringify(selection, null, 2));
        }
    }, [client, tableManagerRef]);
    const handleGoToBottom = react_1.useCallback(() => {
        var _a;
        const size = dataSource ? dataSource.view.size : records.length;
        (_a = tableManagerRef === null || tableManagerRef === void 0 ? void 0 : tableManagerRef.current) === null || _a === void 0 ? void 0 : _a.selectItem(size - 1);
    }, [dataSource, records, tableManagerRef]);
    react_1.useEffect(function setupMenuEntries() {
        if (enableMenuEntries) {
            if (enableClear) {
                client.addMenuEntry({
                    action: 'clear',
                    handler: handleClear,
                });
            }
            if (client.isFB) {
                client.addMenuEntry({
                    action: 'createPaste',
                    handler: handleCreatePaste,
                });
            }
            client.addMenuEntry({
                action: 'goToBottom',
                handler: handleGoToBottom,
            });
        }
    }, [
        client,
        enableClear,
        enableMenuEntries,
        handleClear,
        handleCreatePaste,
        handleGoToBottom,
    ]);
    const table = (React.createElement(DataTable_1.DataTable, { enableAutoScroll: true, ...tableProps, dataSource: dataSource, records: records, tableManagerRef: tableManagerRef, onSelect: handleSelect, extraActions: React.createElement(React.Fragment, null,
            connected && isPaused && (React.createElement(antd_1.Button, { title: `Click to ${pausedState ? 'resume' : 'pause'} the stream`, danger: pausedState, onClick: handleTogglePause }, pausedState ? React.createElement(icons_1.PlayCircleOutlined, null) : React.createElement(icons_1.PauseCircleOutlined, null))),
            connected && enableClear && (React.createElement(antd_1.Button, { title: "Clear records", onClick: handleClear },
                React.createElement(icons_1.DeleteOutlined, null))),
            extraActions) }));
    switch (sidebarPosition) {
        case 'main':
            return (React.createElement(Layout_1.Layout.Container, { grow: true },
                table,
                React.createElement(DetailSidebar_1.DetailSidebar, null, sidebar)));
        case 'right':
            return (React.createElement(Layout_1.Layout.Right, { resizable: true, width: sidebarSize },
                table,
                sidebar));
        case 'bottom':
            return (React.createElement(Layout_1.Layout.Bottom, { resizable: true, height: sidebarSize },
                table,
                sidebar));
        case 'none':
            return table;
    }
}
exports.MasterDetail = MasterDetail;
MasterDetail.defaultProps = {
    sidebarPosition: 'main',
    sidebarSize: 400,
    sidebarComponent: DefaultRenderSidebar,
};
function DefaultRenderSidebar({ record }) {
    return (React.createElement(Panel_1.Panel, { title: "Payload", collapsible: false, pad: true },
        React.createElement(DataInspector_1.DataInspector, { data: record, expandRoot: true })));
}
