// flipper-plugin/lib/ui/NUX.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NUX = exports.NuxManagerContext = exports.createNuxManager = exports.getNuxKey = void 0;
const react_1 = __importStar(require("react"));
const antd_1 = require("antd");
const styled_1 = __importDefault(require("@emotion/styled"));
const css_1 = require("@emotion/css");
const react_element_to_jsx_string_1 = __importDefault(require("react-element-to-jsx-string"));
const PluginContext_1 = require("../plugin/PluginContext");
const atom_1 = require("../state/atom");
const Layout_1 = require("./Layout");
const icons_1 = require("@ant-design/icons");
const crypto_1 = require("crypto");
const theme_1 = require("./theme");
const Tracked_1 = require("./Tracked");
const { Text } = antd_1.Typography;
const storageKey = `FLIPPER_NUX_STATE`;
function getNuxKey(elem, currentPlugin) {
    var _a;
    return `${(_a = currentPlugin === null || currentPlugin === void 0 ? void 0 : currentPlugin.definition.id) !== null && _a !== void 0 ? _a : 'flipper'}:${crypto_1.createHash('sha256')
        .update(react_element_to_jsx_string_1.default(elem))
        .digest('base64')}`;
}
exports.getNuxKey = getNuxKey;
function createNuxManager() {
    const ticker = atom_1.createState(0);
    let readMap = JSON.parse(window.localStorage.getItem(storageKey) || '{}');
    function save() {
        ticker.set(ticker.get() + 1);
        window.localStorage.setItem(storageKey, JSON.stringify(readMap, null, 2));
    }
    return {
        markRead(elem, currentPlugin) {
            readMap[getNuxKey(elem, currentPlugin)] = true;
            save();
        },
        isRead(elem, currentPlugin) {
            return !!readMap[getNuxKey(elem, currentPlugin)];
        },
        resetHints() {
            readMap = {};
            save();
        },
        ticker,
    };
}
exports.createNuxManager = createNuxManager;
const stubManager = {
    markRead() { },
    isRead() {
        return true;
    },
    resetHints() { },
    ticker: atom_1.createState(0),
};
exports.NuxManagerContext = react_1.createContext(stubManager);
function NUX({ children, title, placement, }) {
    const manager = react_1.useContext(exports.NuxManagerContext);
    const pluginInstance = react_1.useContext(PluginContext_1.SandyPluginContext);
    const _tick = atom_1.useValue(manager.ticker);
    const isRead = manager.isRead(title, pluginInstance);
    const dismiss = react_1.useCallback(() => {
        manager.markRead(title, pluginInstance);
    }, [title, manager, pluginInstance]);
    return (react_1.default.createElement(UnanimatedBadge, { count: isRead ? (0) : (react_1.default.createElement(antd_1.Tooltip, { placement: placement, color: theme_1.theme.backgroundWash, title: react_1.default.createElement(Layout_1.Layout.Container, { center: true, gap: true, pad: true, style: { color: theme_1.theme.textColorPrimary } },
                react_1.default.createElement(icons_1.BulbTwoTone, { style: { fontSize: 24 } }),
                react_1.default.createElement(Text, null, title),
                react_1.default.createElement(Tracked_1.Tracked, { action: 'nux:dismiss:' + title.substr(0, 50) },
                    react_1.default.createElement(antd_1.Button, { size: "small", type: "default", onClick: dismiss }, "Dismiss"))) },
            react_1.default.createElement(Pulse, null))) }, children));
}
exports.NUX = NUX;
const UnanimatedBadge = styled_1.default(antd_1.Badge)(({ count }) => ({
    '.ant-scroll-number-custom-component': {
        visibility: count === 0 ? 'hidden' : undefined,
    },
}));
const pulse = css_1.keyframes({
    '0%': {
        opacity: 0.2,
    },
    '100%': {
        opacity: 1,
    },
});
const Pulse = styled_1.default.div({
    cursor: 'pointer',
    background: theme_1.theme.warningColor,
    animation: `${pulse} 2s infinite alternate`,
    borderRadius: 20,
    height: 12,
    width: 12,
    ':hover': {
        opacity: `1 !important`,
        background: theme_1.theme.errorColor,
        animationPlayState: 'paused',
    },
});
