// flipper-plugin/lib/ui/Panel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Panel = void 0;
const React = __importStar(require("react"));
const antd_1 = require("antd");
const Tracked_1 = require("./Tracked");
const useLocalStorageState_1 = require("../utils/useLocalStorageState");
const react_1 = require("react");
const styled_1 = __importDefault(require("@emotion/styled"));
const theme_1 = require("./theme");
const Layout_1 = require("./Layout");
const Panel = (props) => {
    const [collapsed, setCollapsed] = useLocalStorageState_1.useLocalStorageState(`panel:${props.title}:collapsed`, props.collapsible === false ? false : props.collapsed);
    const toggle = react_1.useCallback(() => {
        if (props.collapsible === false) {
            return;
        }
        setCollapsed((c) => !c);
    }, [setCollapsed, props.collapsible]);
    return (React.createElement(Tracked_1.TrackingScope, { scope: props.title },
        React.createElement(StyledCollapse, { bordered: false, activeKey: collapsed ? undefined : props.title, onChange: toggle },
            React.createElement(antd_1.Collapse.Panel, { key: props.title, header: props.extraActions ? (React.createElement(Layout_1.Layout.Right, { center: true },
                    React.createElement("span", null, props.title),
                    props.extraActions)) : (props.title), showArrow: props.collapsible !== false },
                React.createElement(Layout_1.Layout.Container, { pad: props.pad, gap: props.pad }, props.children)))));
};
exports.Panel = Panel;
exports.Panel.defaultProps = {
    collapsed: false,
    collapsible: true,
};
const StyledCollapse = styled_1.default(antd_1.Collapse)({
    background: theme_1.theme.backgroundDefault,
    borderRadius: 0,
    '& > .ant-collapse-item .ant-collapse-header': {
        background: theme_1.theme.backgroundWash,
        paddingTop: theme_1.theme.space.tiny,
        paddingBottom: theme_1.theme.space.tiny,
        paddingLeft: 26,
        fontWeight: 'bold',
        '> .anticon': {
            padding: `5px 0px`,
            left: 8,
            fontSize: '10px',
            fontWeight: theme_1.theme.bold,
        },
    },
    '& > .ant-collapse-item': {
        borderBottom: 'none',
    },
    '& > .ant-collapse-item > .ant-collapse-content > .ant-collapse-content-box': {
        padding: 0,
    },
});
