// flipper-plugin/lib/ui/Sidebar.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sidebar = void 0;
const Layout_1 = require("./Layout");
const react_1 = require("react");
const styled_1 = __importDefault(require("@emotion/styled"));
const react_2 = __importDefault(require("react"));
const icons_1 = require("@ant-design/icons");
const Interactive_1 = require("./Interactive");
const theme_1 = require("./theme");
const SidebarInteractiveContainer = styled_1.default(Interactive_1.Interactive)({
    display: 'flex',
    flex: 1,
});
SidebarInteractiveContainer.displayName = 'Sidebar:SidebarInteractiveContainer';
const borderStyle = `1px solid ${theme_1.theme.dividerColor}`;
const SidebarContainer = styled_1.default(Layout_1.Layout.Container)((props) => ({
    ...(props.unstyled
        ? undefined
        : {
            borderLeft: props.position === 'right' ? borderStyle : 'none',
            borderTop: props.position === 'bottom' ? borderStyle : 'none',
            borderRight: props.position === 'left' ? borderStyle : 'none',
            borderBottom: props.position === 'top' ? borderStyle : 'none',
            backgroundColor: theme_1.theme.backgroundDefault,
        }),
    flex: 1,
    overflow: 'hidden',
}));
SidebarContainer.displayName = 'Sidebar:SidebarContainer';
class Sidebar extends react_1.Component {
    constructor(props, context) {
        super(props, context);
        this.onResize = (width, height) => {
            const { onResize } = this.props;
            if (onResize) {
                onResize(width, height);
            }
            else {
                this.setState({ userChange: true, width, height });
            }
        };
        this.state = {
            userChange: false,
            width: props.width,
            height: props.height,
        };
    }
    static getDerivedStateFromProps(nextProps, state) {
        if (!state.userChange) {
            return { width: nextProps.width, height: nextProps.height };
        }
        return null;
    }
    render() {
        const { onResize, position, children, gutter } = this.props;
        let height;
        let minHeight;
        let maxHeight;
        let width;
        let minWidth;
        let maxWidth;
        const resizable = {};
        if (position === 'left') {
            resizable.right = true;
            ({ width, minWidth, maxWidth } = this.props);
        }
        else if (position === 'top') {
            resizable.bottom = true;
            ({ height, minHeight, maxHeight } = this.props);
        }
        else if (position === 'right') {
            resizable.left = true;
            ({ width, minWidth, maxWidth } = this.props);
        }
        else if (position === 'bottom') {
            resizable.top = true;
            ({ height, minHeight, maxHeight } = this.props);
        }
        const horizontal = position === 'left' || position === 'right';
        const gutterWidth = gutter ? theme_1.theme.space.large : 0;
        if (horizontal) {
            width = width == null ? 200 : width;
            minWidth = (minWidth == null ? 100 : minWidth) + gutterWidth;
            maxWidth = maxWidth == null ? 600 : maxWidth;
        }
        else {
            height = height == null ? 200 : height;
            minHeight = minHeight == null ? 100 : minHeight;
            maxHeight = maxHeight == null ? 600 : maxHeight;
        }
        return (react_2.default.createElement(SidebarInteractiveContainer, { className: this.props.className, minWidth: minWidth, maxWidth: maxWidth, width: horizontal
                ? !children
                    ? gutterWidth
                    : onResize
                        ? width
                        : this.state.width
                : undefined, minHeight: minHeight, maxHeight: maxHeight, height: !horizontal ? (onResize ? height : this.state.height) : undefined, resizable: resizable, onResize: this.onResize, gutterWidth: gutter ? theme_1.theme.space.large : undefined },
            react_2.default.createElement(SidebarContainer, { position: position, unstyled: gutter }, gutter ? (react_2.default.createElement(GutterWrapper, { position: position }, children)) : (children))));
    }
}
exports.Sidebar = Sidebar;
Sidebar.defaultProps = {
    position: 'left',
};
const GutterWrapper = ({ position, children, }) => {
    switch (position) {
        case 'right':
            return (react_2.default.createElement(Layout_1.Layout.Left, null,
                react_2.default.createElement(VerticalGutter, { enabled: !!children }),
                children));
        case 'left':
            return (react_2.default.createElement(Layout_1.Layout.Right, null,
                children,
                react_2.default.createElement(VerticalGutter, { enabled: !!children })));
        case 'bottom':
            return (react_2.default.createElement(Layout_1.Layout.Top, null,
                react_2.default.createElement(VerticalGutter, { enabled: !!children }),
                children));
        case 'top':
            return (react_2.default.createElement(Layout_1.Layout.Bottom, null,
                children,
                react_2.default.createElement(VerticalGutter, { enabled: !!children })));
    }
};
const VerticalGutterContainer = styled_1.default('div')(({ enabled }) => ({
    width: theme_1.theme.space.large,
    minWidth: theme_1.theme.space.large,
    cursor: enabled ? undefined : 'default',
    color: enabled ? theme_1.theme.textColorPlaceholder : theme_1.theme.backgroundWash,
    fontSize: '16px',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    background: theme_1.theme.backgroundWash,
    ':hover': {
        background: enabled ? theme_1.theme.dividerColor : undefined,
    },
}));
const VerticalGutter = ({ enabled }) => (react_2.default.createElement(VerticalGutterContainer, { enabled: enabled },
    react_2.default.createElement(icons_1.MoreOutlined, null)));
