// flipper-plugin/lib/ui/Tabs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tab = exports.Tabs = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const antd_1 = require("antd");
const css_1 = require("@emotion/css");
const Layout_1 = require("./Layout");
const useLocalStorageState_1 = require("../utils/useLocalStorageState");
function Tabs({ grow, children, className, ...baseProps }) {
    const keys = [];
    const keyedChildren = react_1.Children.map(children, (child, idx) => {
        if (!child || typeof child !== 'object') {
            return;
        }
        const tabKey = (typeof child.props.tab === 'string' && child.props.tab) || `tab_${idx}`;
        keys.push(tabKey);
        return {
            ...child,
            props: {
                ...child.props,
                key: tabKey,
                tabKey,
            },
        };
    });
    const [activeTab, setActiveTab] = useLocalStorageState_1.useLocalStorageState('Tabs:' + keys.join(','), undefined);
    return (React.createElement(antd_1.Tabs, { activeKey: activeTab, onChange: (key) => {
            setActiveTab(key);
        }, ...baseProps, className: css_1.cx(className, grow !== false ? growingTabs : undefined) }, keyedChildren));
}
exports.Tabs = Tabs;
const Tab = function Tab({ pad, gap, children, ...baseProps }) {
    return (React.createElement(antd_1.Tabs.TabPane, { ...baseProps },
        React.createElement(Layout_1.Layout.Container, { gap: gap, pad: pad, grow: true }, children)));
};
exports.Tab = Tab;
const growingTabs = css_1.css `
  flex: 1;
  & .tabpanel {
    display: flex;
  }
  & .ant-tabs-content {
    height: 100%;
  }
  & .ant-tabs-tabpane {
    display: flex;
  }
`;
