// flipper-plugin/lib/ui/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeSpace = exports.normalizePadding = exports.theme = void 0;
exports.theme = {
    white: 'white',
    black: 'black',
    primaryColor: 'var(--flipper-primary-color)',
    successColor: 'var(--flipper-success-color)',
    errorColor: 'var(--flipper-error-color)',
    warningColor: 'var(--flipper-warning-color)',
    textColorPrimary: 'var(--flipper-text-color-primary)',
    textColorSecondary: 'var(--flipper-text-color-secondary)',
    textColorPlaceholder: 'var(--flipper-text-color-placeholder)',
    textColorActive: 'var(--light-color-button-active)',
    disabledColor: 'var(--flipper-disabled-color)',
    backgroundDefault: 'var(--flipper-background-default)',
    backgroundWash: 'var(--flipper-background-wash)',
    buttonDefaultBackground: 'var(--flipper-button-default-background)',
    backgroundTransparentHover: 'var(--flipper-background-transparent-hover)',
    dividerColor: 'var(--flipper-divider-color)',
    borderRadius: 'var(--flipper-border-radius)',
    containerBorderRadius: 8,
    inlinePaddingV: 6,
    inlinePaddingH: 12,
    space: {
        tiny: 4,
        small: 8,
        medium: 12,
        large: 16,
        huge: 24,
    },
    fontSize: {
        default: '14px',
        small: '12px',
    },
    monospace: {
        fontFamily: 'SF Mono,Monaco,Andale Mono,monospace',
        fontSize: '12px',
    },
    bold: 600,
};
function normalizePadding({ padv, padh, pad, }) {
    var _a, _b;
    if (padv === undefined && padh === undefined && pad === undefined) {
        return undefined;
    }
    return `${normalizeSpace((_a = padv !== null && padv !== void 0 ? padv : pad) !== null && _a !== void 0 ? _a : 0, exports.theme.inlinePaddingV)}px ${normalizeSpace((_b = padh !== null && padh !== void 0 ? padh : pad) !== null && _b !== void 0 ? _b : 0, exports.theme.inlinePaddingH)}px`;
}
exports.normalizePadding = normalizePadding;
function normalizeSpace(spacing, defaultSpace) {
    return spacing === true
        ? defaultSpace
        : spacing === undefined || spacing === false
            ? 0
            : typeof spacing === 'string'
                ? exports.theme.space[spacing]
                : spacing;
}
exports.normalizeSpace = normalizeSpace;
