// flipper-plugin/lib/ui/Tracked.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withTrackingScope = exports.describeElement = exports.wrapInteractionHandler = exports.useTrackedCallback = exports.Tracked = exports.useCurrentScopeName = exports.TrackingScope = exports.TrackingScopeContext = exports.resetGlobalInteractionReporter = exports.setGlobalInteractionReporter = void 0;
const react_1 = __importStar(require("react"));
const react_2 = require("react");
const react_element_to_jsx_string_1 = __importDefault(require("react-element-to-jsx-string"));
const uuid_1 = require("uuid");
let globalInteractionReporter = () => { };
function setGlobalInteractionReporter(reporter) {
    globalInteractionReporter = reporter;
}
exports.setGlobalInteractionReporter = setGlobalInteractionReporter;
function resetGlobalInteractionReporter() {
    globalInteractionReporter = () => { };
}
exports.resetGlobalInteractionReporter = resetGlobalInteractionReporter;
const DEFAULT_SCOPE = 'Flipper';
exports.TrackingScopeContext = react_2.createContext(DEFAULT_SCOPE);
function TrackingScope({ scope, children, }) {
    const baseScope = react_2.useContext(exports.TrackingScopeContext);
    return (react_1.default.createElement(exports.TrackingScopeContext.Provider, { value: baseScope === DEFAULT_SCOPE ? scope : `${baseScope}:${scope}` }, children));
}
exports.TrackingScope = TrackingScope;
function useCurrentScopeName() {
    return react_2.useContext(exports.TrackingScopeContext);
}
exports.useCurrentScopeName = useCurrentScopeName;
function Tracked({ events = 'onClick', children, action, }) {
    const scope = useCurrentScopeName();
    return react_2.Children.map(children, (child) => {
        if (!child || typeof child !== 'object') {
            return child;
        }
        if (child.type === Tracked)
            return child;
        const newProps = {};
        (typeof events === 'string' ? [events] : events).forEach((event) => {
            const base = child.props[event];
            if (!base) {
                return;
            }
            newProps[event] = wrapInteractionHandler(base, child, event, scope, action);
        });
        return react_2.cloneElement(child, newProps);
    });
}
exports.Tracked = Tracked;
function useTrackedCallback(action, fn, deps) {
    const scope = react_2.useContext(exports.TrackingScopeContext);
    return react_1.useMemo(() => {
        return wrapInteractionHandler(fn, null, '', scope, action);
    }, deps);
}
exports.useTrackedCallback = useTrackedCallback;
function wrapInteractionHandler(fn, element, event, scope, action) {
    const interaction_uuid = uuid_1.v4();
    function report(start, initialEnd, error) {
        const interactionReport = {
            duration: initialEnd - start,
            totalDuration: Date.now() - start,
            success: error ? 0 : 1,
            error: error ? '' + error : undefined,
            componentType: element === null
                ? 'unknown'
                : typeof element === 'string'
                    ? element
                    : describeElementType(element),
            action: action !== null && action !== void 0 ? action : (element && typeof element != 'string'
                ? describeElement(element)
                : 'unknown'),
            scope,
            event,
            uuid: interaction_uuid,
        };
        if (error && typeof error === 'object') {
            error.interaction = interactionReport;
        }
        globalInteractionReporter(interactionReport);
    }
    const res = function trappedInteractionHandler() {
        let res;
        const start = Date.now();
        const r = report.bind(null, start);
        try {
            res = fn.apply(this, arguments);
            if (Date.now() - start > 20) {
                console.warn('Slow interaction');
            }
        }
        catch (e) {
            r(Date.now(), e);
            throw e;
        }
        const initialEnd = Date.now();
        if (typeof (res === null || res === void 0 ? void 0 : res.then) === 'function' && typeof (res === null || res === void 0 ? void 0 : res.catch) === 'function') {
            res.then(() => r(initialEnd), (error) => r(initialEnd, error));
            res = res.catch((error) => {
                return Promise.reject(error);
            });
        }
        else {
            r(initialEnd);
        }
        return res;
    };
    res.flipperTracked = true;
    return res;
}
exports.wrapInteractionHandler = wrapInteractionHandler;
function describeElement(element) {
    var _a;
    const describing = (_a = element.props.title) !== null && _a !== void 0 ? _a : element.props.children;
    if (typeof describing === 'string') {
        return describing;
    }
    if (typeof element.key === 'string') {
        return element.key;
    }
    return stringifyElement(element);
}
exports.describeElement = describeElement;
function describeElementType(element) {
    var _a, _b, _c, _d;
    const t = element.type;
    return typeof t === 'string'
        ? t
        : (_d = (_b = (_a = t === null || t === void 0 ? void 0 : t.displayName) !== null && _a !== void 0 ? _a : t === null || t === void 0 ? void 0 : t.name) !== null && _b !== void 0 ? _b : (_c = t === null || t === void 0 ? void 0 : t.constructor) === null || _c === void 0 ? void 0 : _c.name) !== null && _d !== void 0 ? _d : 'unknown';
}
function withTrackingScope(Component) {
    return function WithTrackingScope(props) {
        var _a, _b, _c;
        const scope = (_b = (_a = Component.displayName) !== null && _a !== void 0 ? _a : Component.name) !== null && _b !== void 0 ? _b : (_c = Component.constructor) === null || _c === void 0 ? void 0 : _c.name;
        if (!scope || typeof scope !== 'string') {
            throw new Error('Failed to find component name for trackingScope');
        }
        return (react_1.default.createElement(TrackingScope, { scope: scope },
            react_1.default.createElement(Component, { ...props })));
    };
}
exports.withTrackingScope = withTrackingScope;
global.FlipperTrackingScopeContext = exports.TrackingScopeContext;
global.FlipperTracked = Tracked;
global.flipperTrackInteraction = function flipperTrackInteraction(elementType, event, scope, action, fn, ...args) {
    if (fn.flipperTracked) {
        return fn(...args);
    }
    return wrapInteractionHandler(fn, elementType, event, scope, !action
        ? 'unknown action'
        : typeof action === 'string'
            ? action
            : stringifyElement(action))(...args);
};
function stringifyElement(element) {
    if (!element)
        return 'unknown element';
    if (typeof element === 'string')
        return element;
    if (Array.isArray(element))
        return element.filter(Boolean).map(stringifyElement).join('');
    return react_element_to_jsx_string_1.default(element).substr(0, 200).replace(/\n/g, ' ');
}
