// flipper-plugin/lib/ui/data-table/ColumnFilter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterButton = exports.FilterIcon = void 0;
const react_1 = require("react");
const styled_1 = __importDefault(require("@emotion/styled"));
const react_2 = __importDefault(require("react"));
const antd_1 = require("antd");
const icons_1 = require("@ant-design/icons");
const theme_1 = require("../theme");
const Layout_1 = require("../Layout");
const { Text } = antd_1.Typography;
function FilterIcon({ column, dispatch, }) {
    const [input, setInput] = react_1.useState('');
    const { filters } = column;
    const isActive = react_1.useMemo(() => filters === null || filters === void 0 ? void 0 : filters.some((f) => f.enabled), [filters]);
    const onAddFilter = (e) => {
        e.stopPropagation();
        dispatch({
            type: 'addColumnFilter',
            column: column.key,
            value: input,
        });
        setInput('');
    };
    const menu = (react_2.default.createElement(antd_1.Menu, { onMouseDown: (e) => {
            e.stopPropagation();
        } },
        react_2.default.createElement(antd_1.Menu.Item, null,
            react_2.default.createElement(Layout_1.Layout.Right, { gap: true },
                react_2.default.createElement(antd_1.Input, { placeholder: "Filter by value", value: input, onChange: (e) => {
                        e.stopPropagation();
                        setInput(e.target.value);
                    }, onClick: (e) => {
                        e.stopPropagation();
                    }, onPressEnter: onAddFilter, disabled: false }),
                react_2.default.createElement(antd_1.Button, { onClick: onAddFilter, title: "Add filter", type: "ghost", style: { padding: '4px 8px' } },
                    react_2.default.createElement(icons_1.PlusCircleOutlined, null)))),
        react_2.default.createElement(antd_1.Menu.Divider, null),
        (filters === null || filters === void 0 ? void 0 : filters.length) ? (filters === null || filters === void 0 ? void 0 : filters.map((filter, index) => (react_2.default.createElement(antd_1.Menu.Item, { key: index },
            react_2.default.createElement(Layout_1.Layout.Right, { center: true },
                react_2.default.createElement(FilterCheckbox, { checked: filter.enabled, onClick: (e) => {
                        e.stopPropagation();
                        e.preventDefault();
                        dispatch({
                            type: 'toggleColumnFilter',
                            column: column.key,
                            index,
                        });
                    } }, filter.label),
                !filter.predefined && (react_2.default.createElement(icons_1.MinusCircleOutlined, { onClick: (e) => {
                        e.stopPropagation();
                        dispatch({
                            type: 'removeColumnFilter',
                            column: column.key,
                            index,
                        });
                    } }))))))) : (react_2.default.createElement(antd_1.Menu.Item, { disabled: true },
            react_2.default.createElement(Text, { type: "secondary", style: { margin: 12 } }, "No active filters"))),
        react_2.default.createElement(antd_1.Menu.Divider, null),
        react_2.default.createElement(antd_1.Menu.Item, { disabled: true },
            react_2.default.createElement("div", { style: { textAlign: 'right' } },
                react_2.default.createElement(antd_1.Button, { type: "link", style: { fontWeight: 'unset' }, onClick: () => {
                        dispatch({
                            type: 'setColumnFilterFromSelection',
                            column: column.key,
                        });
                    } }, "From selection"),
                react_2.default.createElement(antd_1.Button, { type: "link", style: { fontWeight: 'unset' }, onClick: () => {
                        filters === null || filters === void 0 ? void 0 : filters.forEach((f, index) => {
                            if (!f.enabled) {
                                dispatch({
                                    type: 'toggleColumnFilter',
                                    column: column.key,
                                    index,
                                });
                            }
                        });
                    } }, "All"),
                react_2.default.createElement(antd_1.Button, { type: "link", style: { fontWeight: 'unset' }, onClick: () => {
                        filters === null || filters === void 0 ? void 0 : filters.forEach((f, index) => {
                            if (f.enabled)
                                dispatch({
                                    type: 'toggleColumnFilter',
                                    column: column.key,
                                    index,
                                });
                        });
                    } }, "None")))));
    return (react_2.default.createElement(antd_1.Dropdown, { overlay: menu, trigger: ['click'] },
        react_2.default.createElement(exports.FilterButton, { isActive: isActive },
            react_2.default.createElement(icons_1.FilterOutlined, null))));
}
exports.FilterIcon = FilterIcon;
exports.FilterButton = styled_1.default.div(({ isActive }) => ({
    backgroundColor: theme_1.theme.backgroundWash,
    visibility: isActive ? 'visible' : 'hidden',
    color: isActive ? theme_1.theme.textColorActive : theme_1.theme.disabledColor,
    cursor: 'pointer',
    marginRight: 4,
    zIndex: 1,
    '&:hover': {
        color: theme_1.theme.textColorActive,
        backgroundColor: theme_1.theme.backgroundWash,
    },
}));
const FilterCheckbox = styled_1.default(antd_1.Checkbox)({
    maxWidth: 600,
    overflow: 'hidden',
    textOverflow: 'ellipsis',
});
