// flipper-plugin/lib/ui/data-table/TableHead.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableHead = void 0;
const widthUtils_1 = require("../../utils/widthUtils");
const react_1 = require("react");
const Interactive_1 = require("../Interactive");
const styled_1 = __importDefault(require("@emotion/styled"));
const react_2 = __importDefault(require("react"));
const theme_1 = require("../theme");
const antd_1 = require("antd");
const icons_1 = require("@ant-design/icons");
const Layout_1 = require("../Layout");
const ColumnFilter_1 = require("./ColumnFilter");
const toFirstUpper_1 = require("../../utils/toFirstUpper");
const { Text } = antd_1.Typography;
function SortIcons({ direction, onSort, }) {
    return (react_2.default.createElement(SortIconsContainer, { direction: direction },
        react_2.default.createElement(icons_1.CaretUpFilled, { onClick: (e) => {
                e.stopPropagation();
                onSort(direction === 'asc' ? undefined : 'asc');
            }, className: 'ant-table-column-sorter-up ' + (direction === 'asc' ? 'active' : '') }),
        react_2.default.createElement(icons_1.CaretDownFilled, { onClick: (e) => {
                e.stopPropagation();
                onSort(direction === 'desc' ? undefined : 'desc');
            }, className: 'ant-table-column-sorter-down ' +
                (direction === 'desc' ? 'active' : '') })));
}
const SortIconsContainer = styled_1.default.span(({ direction }) => ({
    visibility: direction === undefined ? 'hidden' : undefined,
    display: 'inline-flex',
    flexDirection: 'column',
    alignItems: 'center',
    position: 'relative',
    left: 4,
    top: -3,
    cursor: 'pointer',
    color: theme_1.theme.disabledColor,
    '.ant-table-column-sorter-up:hover, .ant-table-column-sorter-down:hover': {
        color: theme_1.theme.textColorActive,
    },
}));
const TableHeaderColumnInteractive = styled_1.default(Interactive_1.Interactive)({
    overflow: 'hidden',
    whiteSpace: 'nowrap',
    width: '100%',
    borderRight: `1px solid ${theme_1.theme.dividerColor}`,
    paddingRight: 4,
});
TableHeaderColumnInteractive.displayName =
    'TableHead:TableHeaderColumnInteractive';
const TableHeadColumnContainer = styled_1.default.div((props) => ({
    flexShrink: props.width === undefined ? 1 : 0,
    flexGrow: props.width === undefined ? 1 : 0,
    width: props.width === undefined ? '100%' : props.width,
    paddingLeft: 8,
    [`:hover ${SortIconsContainer}`]: {
        visibility: 'visible',
    },
    [`&:hover ${ColumnFilter_1.FilterButton}`]: {
        visibility: 'visible',
    },
}));
TableHeadColumnContainer.displayName = 'TableHead:TableHeadColumnContainer';
const TableHeadContainer = styled_1.default.div(({ scrollbarSize }) => ({
    position: 'relative',
    display: 'flex',
    flexDirection: 'row',
    borderBottom: `1px solid ${theme_1.theme.dividerColor}`,
    backgroundColor: theme_1.theme.backgroundWash,
    userSelect: 'none',
    whiteSpace: 'nowrap',
    borderLeft: `4px solid ${theme_1.theme.backgroundWash}`,
    paddingRight: scrollbarSize,
}));
TableHeadContainer.displayName = 'TableHead:TableHeadContainer';
const RIGHT_RESIZABLE = { right: true };
function TableHeadColumn({ column, isResizable, sorted, dispatch, }) {
    const ref = react_1.useRef(null);
    const onResize = (newWidth) => {
        if (!isResizable) {
            return;
        }
        let normalizedWidth = newWidth;
        if (widthUtils_1.isPercentage(column.width) && ref.current) {
            const { parentElement } = ref.current;
            const parentWidth = parentElement.clientWidth;
            const { childNodes } = parentElement;
            const lastElem = childNodes[childNodes.length - 1];
            const right = lastElem instanceof HTMLElement
                ? lastElem.offsetLeft + lastElem.clientWidth + 1
                : 0;
            if (right < parentWidth) {
                normalizedWidth = widthUtils_1.calculatePercentage(parentWidth, newWidth);
            }
        }
        dispatch({
            type: 'resizeColumn',
            column: column.key,
            width: normalizedWidth,
        });
    };
    let children = (react_2.default.createElement(Layout_1.Layout.Right, { center: true },
        react_2.default.createElement("div", { onClick: (e) => {
                e.stopPropagation();
                const newDirection = sorted === undefined
                    ? 'asc'
                    : sorted === 'asc'
                        ? 'desc'
                        : undefined;
                dispatch({
                    type: 'sortColumn',
                    column: column.key,
                    direction: newDirection,
                });
            }, role: "button", tabIndex: 0 },
            react_2.default.createElement(Text, { type: "secondary" },
                column.title === undefined ? (toFirstUpper_1.toFirstUpper(column.key)) : column.title === '' ? (react_2.default.createElement(react_2.default.Fragment, null, "\u00A0")) : (column.title),
                react_2.default.createElement(SortIcons, { direction: sorted, onSort: (dir) => dispatch({ type: 'sortColumn', column: column.key, direction: dir }) }))),
        react_2.default.createElement(ColumnFilter_1.FilterIcon, { column: column, dispatch: dispatch })));
    if (isResizable) {
        children = (react_2.default.createElement(TableHeaderColumnInteractive, { grow: true, resizable: RIGHT_RESIZABLE, onResize: onResize, minWidth: 20 }, children));
    }
    return (react_2.default.createElement(TableHeadColumnContainer, { width: column.width, ref: ref }, children));
}
exports.TableHead = react_1.memo(function TableHead({ visibleColumns, dispatch, sorting, scrollbarSize, }) {
    return (react_2.default.createElement(TableHeadContainer, { scrollbarSize: scrollbarSize }, visibleColumns.map((column, i) => (react_2.default.createElement(TableHeadColumn, { key: column.key, column: column, isResizable: i < visibleColumns.length - 1, dispatch: dispatch, sorted: (sorting === null || sorting === void 0 ? void 0 : sorting.key) === column.key ? sorting.direction : undefined })))));
});
