// flipper-plugin/lib/ui/data-table/TableContextMenu.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tableContextMenuFactory = void 0;
const icons_1 = require("@ant-design/icons");
const antd_1 = require("antd");
const DataTableManager_1 = require("./DataTableManager");
const react_1 = __importDefault(require("react"));
const FlipperLib_1 = require("../../plugin/FlipperLib");
const toFirstUpper_1 = require("../../utils/toFirstUpper");
const { Item, SubMenu } = antd_1.Menu;
function defaultOnCopyRows(items) {
    return JSON.stringify(items.length > 1 ? items : items[0], null, 2);
}
function tableContextMenuFactory(datasource, dispatch, selection, columns, visibleColumns, onCopyRows = defaultOnCopyRows, onContextMenu) {
    var _a;
    const lib = FlipperLib_1.tryGetFlipperLibImplementation();
    if (!lib) {
        return (react_1.default.createElement(antd_1.Menu, null,
            react_1.default.createElement(Item, null, "Menu not ready")));
    }
    const hasSelection = (_a = selection.items.size > 0) !== null && _a !== void 0 ? _a : false;
    return (react_1.default.createElement(antd_1.Menu, null,
        onContextMenu
            ? onContextMenu(DataTableManager_1.getSelectedItem(datasource, selection))
            : null,
        react_1.default.createElement(SubMenu, { key: "filter same", title: "Filter on same", icon: react_1.default.createElement(icons_1.FilterOutlined, null), disabled: !hasSelection }, visibleColumns.map((column, idx) => {
            var _a;
            return (react_1.default.createElement(Item, { key: (_a = column.key) !== null && _a !== void 0 ? _a : idx, onClick: () => {
                    dispatch({
                        type: 'setColumnFilterFromSelection',
                        column: column.key,
                    });
                } }, friendlyColumnTitle(column)));
        })),
        react_1.default.createElement(SubMenu, { key: "copy cells", title: "Copy cell(s)", icon: react_1.default.createElement(icons_1.CopyOutlined, null), disabled: !hasSelection }, visibleColumns.map((column, idx) => {
            var _a;
            return (react_1.default.createElement(Item, { key: (_a = column.key) !== null && _a !== void 0 ? _a : idx, onClick: () => {
                    const items = DataTableManager_1.getSelectedItems(datasource, selection);
                    if (items.length) {
                        lib.writeTextToClipboard(items.map((item) => '' + item[column.key]).join('\n'));
                    }
                } }, friendlyColumnTitle(column)));
        })),
        react_1.default.createElement(Item, { key: "copyToClipboard", disabled: !hasSelection, onClick: () => {
                const items = DataTableManager_1.getSelectedItems(datasource, selection);
                if (items.length) {
                    lib.writeTextToClipboard(onCopyRows(items));
                }
            } }, "Copy row(s)"),
        lib.isFB && (react_1.default.createElement(Item, { key: "createPaste", disabled: !hasSelection, onClick: () => {
                const items = DataTableManager_1.getSelectedItems(datasource, selection);
                if (items.length) {
                    lib.createPaste(onCopyRows(items));
                }
            } }, "Create paste")),
        react_1.default.createElement(antd_1.Menu.Divider, null),
        react_1.default.createElement(SubMenu, { title: "Visible columns", key: "visible columns" }, columns.map((column, idx) => {
            var _a;
            return (react_1.default.createElement(antd_1.Menu.Item, { key: (_a = column.key) !== null && _a !== void 0 ? _a : idx },
                react_1.default.createElement(antd_1.Checkbox, { checked: column.visible, onClick: (e) => {
                        e.stopPropagation();
                        e.preventDefault();
                        dispatch({ type: 'toggleColumnVisibility', column: column.key });
                    } }, friendlyColumnTitle(column))));
        })),
        react_1.default.createElement(antd_1.Menu.Item, { key: "reset", onClick: () => {
                dispatch({ type: 'reset' });
            } }, "Reset view")));
}
exports.tableContextMenuFactory = tableContextMenuFactory;
function friendlyColumnTitle(column) {
    const name = column.title || column.key;
    return toFirstUpper_1.toFirstUpper(name);
}
