// flipper-plugin/lib/ui/data-table/TableSearch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableSearch = void 0;
const icons_1 = require("@ant-design/icons");
const antd_1 = require("antd");
const react_1 = __importStar(require("react"));
const styled_1 = __importDefault(require("@emotion/styled"));
const Layout_1 = require("../Layout");
const theme_1 = require("../theme");
exports.TableSearch = react_1.memo(function TableSearch({ searchValue, useRegex, dispatch, extraActions, contextMenu, }) {
    const onSearch = react_1.useCallback((value) => {
        dispatch({ type: 'setSearchValue', value });
    }, [dispatch]);
    const onToggleRegex = react_1.useCallback((e) => {
        e.stopPropagation();
        e.preventDefault();
        dispatch({ type: 'toggleUseRegex' });
    }, [dispatch]);
    const regexError = react_1.useMemo(() => {
        if (!useRegex || !searchValue) {
            return;
        }
        try {
            new RegExp(searchValue);
        }
        catch (e) {
            return '' + e;
        }
    }, [useRegex, searchValue]);
    return (react_1.default.createElement(Searchbar, { gap: true },
        react_1.default.createElement(antd_1.Input.Search, { allowClear: true, placeholder: "Search...", onSearch: onSearch, value: searchValue, suffix: react_1.default.createElement(RegexButton, { size: "small", onClick: onToggleRegex, style: useRegex
                    ? {
                        background: regexError
                            ? theme_1.theme.errorColor
                            : theme_1.theme.successColor,
                        color: theme_1.theme.white,
                    }
                    : {
                        color: theme_1.theme.disabledColor,
                    }, type: "default", title: regexError || 'Search using Regex' }, ".*"), onChange: (e) => {
                onSearch(e.target.value);
            } }),
        extraActions,
        contextMenu && (react_1.default.createElement(antd_1.Dropdown, { overlay: contextMenu, placement: "bottomRight" },
            react_1.default.createElement(antd_1.Button, { type: "text", size: "small", style: { height: '100%' } },
                react_1.default.createElement(icons_1.MenuOutlined, null))))));
});
const Searchbar = styled_1.default(Layout_1.Layout.Horizontal)({
    backgroundColor: theme_1.theme.backgroundWash,
    padding: theme_1.theme.space.small,
    '.ant-input-affix-wrapper': {
        height: 32,
    },
    '.ant-btn': {
        padding: `${theme_1.theme.space.tiny}px ${theme_1.theme.space.small}px`,
        background: 'transparent',
    },
});
const RegexButton = styled_1.default(antd_1.Button)({
    padding: '0px !important',
    borderRadius: 4,
    marginRight: -6,
    marginLeft: 4,
    lineHeight: '20px',
    width: 20,
    height: 20,
    border: 'none',
    '& :hover': {
        color: theme_1.theme.primaryColor,
    },
});
