// flipper-plugin/lib/utils/LowPassFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class LowPassFilter {
    constructor(smoothing = 0.9) {
        this._nextReduce = (last, current) => {
            return this.smoothing * current + (1 - this.smoothing) * last;
        };
        this.smoothing = smoothing;
        this.buffer = [];
        this.bufferMaxSize = 5;
    }
    hasFullBuffer() {
        return this.buffer.length === this.bufferMaxSize;
    }
    push(value) {
        let removed = 0;
        if (this.hasFullBuffer()) {
            const tmp = this.buffer.shift();
            if (tmp === undefined)
                throw new Error('Invariant violation: Buffer reported full but shift returned nothing.');
            removed = tmp;
        }
        this.buffer.push(value);
        return removed;
    }
    next(nextValue) {
        const removed = this.push(nextValue);
        const result = this.buffer.reduce(this._nextReduce, removed);
        this.buffer[this.buffer.length - 1] = result;
        return result;
    }
}
exports.default = LowPassFilter;
