// flipper-plugin/lib/plugin/DevicePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandyDevicePluginInstance = void 0;
const PluginBase_1 = require("./PluginBase");
class SandyDevicePluginInstance extends PluginBase_1.BasePluginInstance {
    constructor(flipperLib, definition, realDevice, pluginKey, initialStates) {
        super(flipperLib, definition, realDevice, pluginKey, initialStates);
        this.client = {
            ...this.createBasePluginClient(),
            isPluginAvailable(pluginId) {
                return flipperLib.isPluginAvailable(realDevice, null, pluginId);
            },
            selectPlugin(pluginId, deeplink) {
                if (this.isPluginAvailable(pluginId)) {
                    flipperLib.selectPlugin(realDevice, null, pluginId, deeplink);
                }
            },
            get isConnected() {
                return realDevice.connected.get();
            },
            connected: realDevice.connected,
        };
        this.initializePlugin(() => definition.asDevicePluginModule().devicePlugin(this.client));
    }
    static is(thing) {
        return thing instanceof SandyDevicePluginInstance;
    }
    toJSON() {
        return '[SandyDevicePluginInstance]';
    }
}
exports.SandyDevicePluginInstance = SandyDevicePluginInstance;
