// flipper-plugin/lib/plugin/SandyPluginDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandyPluginDefinition = void 0;
class SandyPluginDefinition {
    constructor(details, module) {
        this.id = details.id;
        this.details = details;
        if (details.pluginType === 'device' ||
            module.supportsDevice ||
            module.devicePlugin) {
            this.isDevicePlugin = true;
            if (!module.devicePlugin || typeof module.devicePlugin !== 'function') {
                throw new Error(`Flipper device plugin '${this.id}' should export named function called 'devicePlugin'`);
            }
        }
        else {
            this.isDevicePlugin = false;
            if (!module.plugin || typeof module.plugin !== 'function') {
                throw new Error(`Flipper plugin '${this.id}' should export named function called 'plugin'`);
            }
        }
        if (!module.Component || typeof module.Component !== 'function') {
            throw new Error(`Flipper plugin '${this.id}' should export named function called 'Component'`);
        }
        this.module = module;
        this.module.Component.displayName = `FlipperPlugin(${this.id})`;
    }
    asDevicePluginModule() {
        if (!this.isDevicePlugin)
            throw new Error('Not a device plugin');
        return this.module;
    }
    asPluginModule() {
        if (this.isDevicePlugin)
            throw new Error('Not an application plugin');
        return this.module;
    }
    get packageName() {
        return this.details.name;
    }
    get title() {
        return this.details.title;
    }
    get icon() {
        return this.details.icon;
    }
    get category() {
        return this.details.category;
    }
    get gatekeeper() {
        return this.details.gatekeeper;
    }
    get version() {
        return this.details.version;
    }
    get isBundled() {
        return this.details.isBundled;
    }
    get keyboardActions() {
        return [];
    }
}
exports.SandyPluginDefinition = SandyPluginDefinition;
