// flipper-plugin/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTablePlugin = exports.useMemoize = exports.ElementsInspector = exports.DataInspector = exports.MarkerTimeline = exports.DataDescription = exports.useLocalStorageState = exports.Tab = exports.Tabs = exports.Panel = exports._Interactive = exports.DataList = exports.DataTable = exports.createDataSource = exports.DataSource = exports._LoggerContext = exports.useLogger = exports.sleep = exports.DataFormatter = exports._wrapInteractionHandler = exports.useTrackedCallback = exports.withTrackingScope = exports._setGlobalInteractionReporter = exports.TrackingScope = exports.Tracked = exports.renderReactRoot = exports.CodeBlock = exports.MasterDetail = exports.Toolbar = exports.DetailSidebar = exports._Sidebar = exports._createNuxManager = exports._NuxManagerContext = exports.NUX = exports.Layout = exports.theme = exports._buildInMenuEntries = exports._setFlipperLibImplementation = exports._getFlipperLibImplementation = exports.batch = exports.useValue = exports.createState = exports.usePlugin = exports._SandyPluginContext = exports._SandyPluginRenderer = exports._SandyPluginDefinition = exports._SandyDevicePluginInstance = exports._SandyPluginInstance = exports.styled = exports.produce = void 0;
exports.TestUtils = void 0;
var immer_1 = require("immer");
Object.defineProperty(exports, "produce", { enumerable: true, get: function () { return immer_1.produce; } });
const styled_1 = __importDefault(require("@emotion/styled"));
exports.styled = styled_1.default;
require("./plugin/PluginBase");
const TestUtilites = __importStar(require("./test-utils/test-utils"));
var Plugin_1 = require("./plugin/Plugin");
Object.defineProperty(exports, "_SandyPluginInstance", { enumerable: true, get: function () { return Plugin_1.SandyPluginInstance; } });
var DevicePlugin_1 = require("./plugin/DevicePlugin");
Object.defineProperty(exports, "_SandyDevicePluginInstance", { enumerable: true, get: function () { return DevicePlugin_1.SandyDevicePluginInstance; } });
var SandyPluginDefinition_1 = require("./plugin/SandyPluginDefinition");
Object.defineProperty(exports, "_SandyPluginDefinition", { enumerable: true, get: function () { return SandyPluginDefinition_1.SandyPluginDefinition; } });
var PluginRenderer_1 = require("./plugin/PluginRenderer");
Object.defineProperty(exports, "_SandyPluginRenderer", { enumerable: true, get: function () { return PluginRenderer_1.SandyPluginRenderer; } });
var PluginContext_1 = require("./plugin/PluginContext");
Object.defineProperty(exports, "_SandyPluginContext", { enumerable: true, get: function () { return PluginContext_1.SandyPluginContext; } });
Object.defineProperty(exports, "usePlugin", { enumerable: true, get: function () { return PluginContext_1.usePlugin; } });
var atom_1 = require("./state/atom");
Object.defineProperty(exports, "createState", { enumerable: true, get: function () { return atom_1.createState; } });
Object.defineProperty(exports, "useValue", { enumerable: true, get: function () { return atom_1.useValue; } });
var batch_1 = require("./state/batch");
Object.defineProperty(exports, "batch", { enumerable: true, get: function () { return batch_1.batch; } });
var FlipperLib_1 = require("./plugin/FlipperLib");
Object.defineProperty(exports, "_getFlipperLibImplementation", { enumerable: true, get: function () { return FlipperLib_1.getFlipperLibImplementation; } });
Object.defineProperty(exports, "_setFlipperLibImplementation", { enumerable: true, get: function () { return FlipperLib_1.setFlipperLibImplementation; } });
var MenuEntry_1 = require("./plugin/MenuEntry");
Object.defineProperty(exports, "_buildInMenuEntries", { enumerable: true, get: function () { return MenuEntry_1.buildInMenuEntries; } });
var theme_1 = require("./ui/theme");
Object.defineProperty(exports, "theme", { enumerable: true, get: function () { return theme_1.theme; } });
var Layout_1 = require("./ui/Layout");
Object.defineProperty(exports, "Layout", { enumerable: true, get: function () { return Layout_1.Layout; } });
var NUX_1 = require("./ui/NUX");
Object.defineProperty(exports, "NUX", { enumerable: true, get: function () { return NUX_1.NUX; } });
Object.defineProperty(exports, "_NuxManagerContext", { enumerable: true, get: function () { return NUX_1.NuxManagerContext; } });
Object.defineProperty(exports, "_createNuxManager", { enumerable: true, get: function () { return NUX_1.createNuxManager; } });
var Sidebar_1 = require("./ui/Sidebar");
Object.defineProperty(exports, "_Sidebar", { enumerable: true, get: function () { return Sidebar_1.Sidebar; } });
var DetailSidebar_1 = require("./ui/DetailSidebar");
Object.defineProperty(exports, "DetailSidebar", { enumerable: true, get: function () { return DetailSidebar_1.DetailSidebar; } });
var Toolbar_1 = require("./ui/Toolbar");
Object.defineProperty(exports, "Toolbar", { enumerable: true, get: function () { return Toolbar_1.Toolbar; } });
var MasterDetail_1 = require("./ui/MasterDetail");
Object.defineProperty(exports, "MasterDetail", { enumerable: true, get: function () { return MasterDetail_1.MasterDetail; } });
var CodeBlock_1 = require("./ui/CodeBlock");
Object.defineProperty(exports, "CodeBlock", { enumerable: true, get: function () { return CodeBlock_1.CodeBlock; } });
var renderReactRoot_1 = require("./utils/renderReactRoot");
Object.defineProperty(exports, "renderReactRoot", { enumerable: true, get: function () { return renderReactRoot_1.renderReactRoot; } });
var Tracked_1 = require("./ui/Tracked");
Object.defineProperty(exports, "Tracked", { enumerable: true, get: function () { return Tracked_1.Tracked; } });
Object.defineProperty(exports, "TrackingScope", { enumerable: true, get: function () { return Tracked_1.TrackingScope; } });
Object.defineProperty(exports, "_setGlobalInteractionReporter", { enumerable: true, get: function () { return Tracked_1.setGlobalInteractionReporter; } });
Object.defineProperty(exports, "withTrackingScope", { enumerable: true, get: function () { return Tracked_1.withTrackingScope; } });
Object.defineProperty(exports, "useTrackedCallback", { enumerable: true, get: function () { return Tracked_1.useTrackedCallback; } });
Object.defineProperty(exports, "_wrapInteractionHandler", { enumerable: true, get: function () { return Tracked_1.wrapInteractionHandler; } });
var DataFormatter_1 = require("./ui/DataFormatter");
Object.defineProperty(exports, "DataFormatter", { enumerable: true, get: function () { return DataFormatter_1.DataFormatter; } });
var sleep_1 = require("./utils/sleep");
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return sleep_1.sleep; } });
var Logger_1 = require("./utils/Logger");
Object.defineProperty(exports, "useLogger", { enumerable: true, get: function () { return Logger_1.useLogger; } });
Object.defineProperty(exports, "_LoggerContext", { enumerable: true, get: function () { return Logger_1._LoggerContext; } });
var index_1 = require("./data-source/index");
Object.defineProperty(exports, "DataSource", { enumerable: true, get: function () { return index_1.DataSource; } });
var createDataSource_1 = require("./state/createDataSource");
Object.defineProperty(exports, "createDataSource", { enumerable: true, get: function () { return createDataSource_1.createDataSource; } });
var DataTable_1 = require("./ui/data-table/DataTable");
Object.defineProperty(exports, "DataTable", { enumerable: true, get: function () { return DataTable_1.DataTable; } });
var DataList_1 = require("./ui/DataList");
Object.defineProperty(exports, "DataList", { enumerable: true, get: function () { return DataList_1.DataList; } });
var Interactive_1 = require("./ui/Interactive");
Object.defineProperty(exports, "_Interactive", { enumerable: true, get: function () { return Interactive_1.Interactive; } });
var Panel_1 = require("./ui/Panel");
Object.defineProperty(exports, "Panel", { enumerable: true, get: function () { return Panel_1.Panel; } });
var Tabs_1 = require("./ui/Tabs");
Object.defineProperty(exports, "Tabs", { enumerable: true, get: function () { return Tabs_1.Tabs; } });
Object.defineProperty(exports, "Tab", { enumerable: true, get: function () { return Tabs_1.Tab; } });
var useLocalStorageState_1 = require("./utils/useLocalStorageState");
Object.defineProperty(exports, "useLocalStorageState", { enumerable: true, get: function () { return useLocalStorageState_1.useLocalStorageState; } });
var DataDescription_1 = require("./ui/data-inspector/DataDescription");
Object.defineProperty(exports, "DataDescription", { enumerable: true, get: function () { return DataDescription_1.DataDescription; } });
var MarkerTimeline_1 = require("./ui/MarkerTimeline");
Object.defineProperty(exports, "MarkerTimeline", { enumerable: true, get: function () { return MarkerTimeline_1.MarkerTimeline; } });
var DataInspector_1 = require("./ui/data-inspector/DataInspector");
Object.defineProperty(exports, "DataInspector", { enumerable: true, get: function () { return DataInspector_1.DataInspector; } });
var ElementsInspector_1 = require("./ui/elements-inspector/ElementsInspector");
Object.defineProperty(exports, "ElementsInspector", { enumerable: true, get: function () { return ElementsInspector_1.ElementsInspector; } });
var useMemoize_1 = require("./utils/useMemoize");
Object.defineProperty(exports, "useMemoize", { enumerable: true, get: function () { return useMemoize_1.useMemoize; } });
var createTablePlugin_1 = require("./utils/createTablePlugin");
Object.defineProperty(exports, "createTablePlugin", { enumerable: true, get: function () { return createTablePlugin_1.createTablePlugin; } });
exports.TestUtils = TestUtilites;
