// flipper-plugin/lib/ui/data-inspector/DataPreview.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InspectorName = void 0;
const DataDescription_1 = require("./DataDescription");
const styled_1 = __importDefault(require("@emotion/styled"));
const utils_1 = require("./utils");
const react_1 = require("react");
const react_2 = __importDefault(require("react"));
exports.InspectorName = styled_1.default.span({
    color: '#7b64c0',
});
exports.InspectorName.displayName = 'DataInspector:InspectorName';
const PreviewContainer = styled_1.default.span({
    fontStyle: 'italic',
});
PreviewContainer.displayName = 'DataPreview:PreviewContainer';
function intersperse(arr, sep) {
    if (arr.length === 0) {
        return [];
    }
    return arr.slice(1).reduce((xs, x) => {
        return xs.concat([sep, x]);
    }, [arr[0]]);
}
class DataPreview extends react_1.PureComponent {
    render() {
        const { depth, extractValue, path, type, value } = this.props;
        if (type === 'array') {
            return (react_2.default.createElement(PreviewContainer, null,
                '[',
                intersperse(value.map((element, index) => {
                    const res = extractValue(element, depth + 1, path);
                    if (!res) {
                        return null;
                    }
                    const { type, value } = res;
                    return (react_2.default.createElement(DataDescription_1.DataDescription, { key: index, type: type, value: value, setValue: null }));
                }), ', '),
                ']'));
        }
        else if (type === 'date') {
            return react_2.default.createElement("span", null, value.toString());
        }
        else if (type === 'object') {
            const propertyNodes = [];
            const keys = utils_1.getSortedKeys(value);
            let i = 0;
            for (const key of keys) {
                let ellipsis;
                i++;
                if (i >= this.props.maxProperties) {
                    ellipsis = react_2.default.createElement("span", { key: 'ellipsis' }, "\u2026");
                }
                propertyNodes.push(react_2.default.createElement("span", { key: key },
                    react_2.default.createElement(exports.InspectorName, null, key),
                    ellipsis));
                if (ellipsis) {
                    break;
                }
            }
            return (react_2.default.createElement(PreviewContainer, null,
                '{',
                intersperse(propertyNodes, ', '),
                '}'));
        }
        else {
            return null;
        }
    }
}
exports.default = DataPreview;
DataPreview.defaultProps = {
    maxProperties: 5,
};
