// flipper-plugin/lib/ui/data-inspector/DataInspector.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataInspector = void 0;
const DataInspectorNode_1 = require("./DataInspectorNode");
const react_1 = require("react");
const DataInspectorNode_2 = require("./DataInspectorNode");
const react_2 = __importDefault(require("react"));
const Highlight_1 = require("../Highlight");
const MAX_RESULTS = 50;
const EMPTY_ARRAY = [];
class DataInspector extends react_1.PureComponent {
    constructor() {
        super(...arguments);
        this.state = {
            expanded: {},
            userExpanded: {},
            filterExpanded: {},
            filter: '',
        };
        this.onExpanded = (path, isExpanded) => {
            this.setState({
                userExpanded: {
                    ...this.state.userExpanded,
                    [path]: isExpanded,
                },
                expanded: {
                    ...this.state.expanded,
                    [path]: isExpanded,
                },
            });
        };
        this.getRootData = () => {
            return this.props.data;
        };
    }
    static getDerivedStateFromProps(nextProps, currentState) {
        var _a;
        if (((_a = nextProps.filter) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === currentState.filter) {
            return null;
        }
        if (!nextProps.filter) {
            return {
                filter: '',
                filterExpanded: {},
                expanded: currentState.userExpanded,
            };
        }
        const filter = nextProps.filter.toLowerCase();
        const paths = [];
        function walk(value, path) {
            if (paths.length > MAX_RESULTS) {
                return;
            }
            if (!value) {
                return;
            }
            if (typeof value !== 'object') {
                if (('' + value).toLowerCase().includes(filter)) {
                    paths.push(path.slice());
                }
            }
            else if (Array.isArray(value)) {
                value.forEach((value, index) => {
                    path.push(index);
                    walk(value, path);
                    path.pop();
                });
            }
            else {
                Object.keys(value).forEach((key) => {
                    path.push(key);
                    walk(key, path);
                    walk(value[key], path);
                    path.pop();
                });
            }
        }
        if (filter.length >= 2) {
            walk(nextProps.data, []);
        }
        const filterExpanded = {};
        paths.forEach((path) => {
            for (let i = 1; i < path.length; i++)
                filterExpanded[path.slice(0, i).join('.')] = true;
        });
        return {
            filterExpanded,
            expanded: { ...currentState.userExpanded, ...filterExpanded },
            filter,
        };
    }
    render() {
        return (react_2.default.createElement(DataInspectorNode_1.RootDataContext.Provider, { value: this.getRootData },
            react_2.default.createElement(Highlight_1.HighlightProvider, { text: this.props.filter },
                react_2.default.createElement(DataInspectorNode_2.DataInspectorNode, { data: this.props.data, diff: this.props.diff, extractValue: this.props.extractValue, setValue: this.props.setValue, expanded: this.state.expanded, onExpanded: this.onExpanded, onDelete: this.props.onDelete, onRenderName: this.props.onRenderName, onRenderDescription: this.props.onRenderDescription, expandRoot: this.props.expandRoot, collapsed: this.props.filter ? true : this.props.collapsed, tooltips: this.props.tooltips, parentPath: EMPTY_ARRAY, depth: 0, parentAncestry: EMPTY_ARRAY }))));
    }
}
exports.DataInspector = DataInspector;
