// flipper-plugin/lib/state/atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAtom = exports.useValue = exports.createState = void 0;
const immer_1 = require("immer");
const react_1 = require("react");
const PluginBase_1 = require("../plugin/PluginBase");
immer_1.enableMapSet();
class AtomValue {
    constructor(initialValue) {
        this.listeners = [];
        this.value = initialValue;
    }
    get() {
        return this.value;
    }
    set(nextValue) {
        if (nextValue !== this.value) {
            const prevValue = this.value;
            this.value = nextValue;
            this.notifyChanged(prevValue);
        }
    }
    deserialize(value) {
        this.set(value);
    }
    serialize() {
        return this.get();
    }
    update(recipe) {
        this.set(immer_1.produce(this.value, recipe));
    }
    notifyChanged(prevValue) {
        this.listeners.slice().forEach((l) => l(this.value, prevValue));
    }
    subscribe(listener) {
        this.listeners.push(listener);
        return () => this.unsubscribe(listener);
    }
    unsubscribe(listener) {
        const idx = this.listeners.indexOf(listener);
        if (idx !== -1) {
            this.listeners.splice(idx, 1);
        }
    }
}
function createState(initialValue = undefined, options = {}) {
    const atom = new AtomValue(initialValue);
    PluginBase_1.registerStorageAtom(options.persist, atom);
    return atom;
}
exports.createState = createState;
function useValue(atom, defaultValue) {
    const [localValue, setLocalValue] = react_1.useState(atom ? atom.get() : defaultValue);
    react_1.useEffect(() => {
        if (!atom) {
            return;
        }
        setLocalValue(atom.get());
        atom.subscribe(setLocalValue);
        return () => {
            atom.unsubscribe(setLocalValue);
        };
    }, [atom]);
    return localValue;
}
exports.useValue = useValue;
function isAtom(value) {
    return value instanceof AtomValue;
}
exports.isAtom = isAtom;
