// flipper-client-sdk/lib/webviewImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.newWebviewClient = void 0;
const api_1 = require("./api");
class FlipperWebviewClient extends api_1.FlipperClient {
    constructor() {
        super(...arguments);
        this._subscriptions = new Map();
        this._client = null;
        this.start = (appName) => {
            const bridge = window.FlipperWebviewBridge;
            bridge === null || bridge === void 0 ? void 0 : bridge.registerPlugins(this.plugins);
            bridge === null || bridge === void 0 ? void 0 : bridge.start(appName);
        };
        this.stop = () => {
            const bridge = window.FlipperWebviewBridge;
            bridge === null || bridge === void 0 ? void 0 : bridge.FlipperWebviewBridge.stop();
        };
        this.sendData = (data) => {
            const bridge = window.FlipperWebviewBridge;
            bridge && bridge.sendFlipperObject(data);
        };
        this.isAvailable = () => {
            return window.FlipperWebviewBridge != null;
        };
    }
}
function newWebviewClient() {
    return new FlipperWebviewClient();
}
exports.newWebviewClient = newWebviewClient;
