// flipper-plugin-lib/lib/getPluginDetails.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDownloadablePluginDetails = exports.getInstalledPluginDetails = exports.getPluginDetails = exports.isPluginDir = exports.isPluginJson = exports.readPluginPackageJson = void 0;
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
const pluginPaths_1 = require("./pluginPaths");
async function readPluginPackageJson(dir) {
    const baseJson = await fs_extra_1.default.readJson(path_1.default.join(dir, 'package.json'));
    if (await fs_extra_1.default.pathExists(path_1.default.join(dir, 'fb', 'package.json'))) {
        const addedJson = await fs_extra_1.default.readJson(path_1.default.join(dir, 'fb', 'package.json'));
        return Object.assign({}, baseJson, addedJson);
    }
    else {
        return baseJson;
    }
}
exports.readPluginPackageJson = readPluginPackageJson;
function isPluginJson(packageJson) {
    var _a;
    return (_a = packageJson === null || packageJson === void 0 ? void 0 : packageJson.keywords) === null || _a === void 0 ? void 0 : _a.includes('flipper-plugin');
}
exports.isPluginJson = isPluginJson;
async function isPluginDir(dir) {
    const packageJsonPath = path_1.default.join(dir, 'package.json');
    const json = (await fs_extra_1.default.pathExists(packageJsonPath))
        ? await fs_extra_1.default.readJson(path_1.default.join(dir, 'package.json'), {
            throws: false,
        })
        : undefined;
    return isPluginJson(json);
}
exports.isPluginDir = isPluginDir;
function getPluginDetails(packageJson) {
    const specVersion = packageJson.$schema &&
        packageJson.$schema ===
            'https://fbflipper.com/schemas/plugin-package/v2.json'
        ? 2
        : 1;
    switch (specVersion) {
        case 1:
            return getPluginDetailsV1(packageJson);
        case 2:
            return getPluginDetailsV2(packageJson);
        default:
            throw new Error(`Unknown plugin format version: ${specVersion}`);
    }
}
exports.getPluginDetails = getPluginDetails;
async function getInstalledPluginDetails(dir, packageJson) {
    packageJson = packageJson !== null && packageJson !== void 0 ? packageJson : (await readPluginPackageJson(dir));
    const pluginDetails = getPluginDetails(packageJson);
    const entry = pluginDetails.specVersion === 1
        ? path_1.default.resolve(pluginPaths_1.pluginCacheDir, `${packageJson.name}@${packageJson.version || '0.0.0'}.js`)
        : path_1.default.resolve(dir, packageJson.main);
    return {
        ...pluginDetails,
        isBundled: false,
        isActivatable: true,
        dir,
        entry,
    };
}
exports.getInstalledPluginDetails = getInstalledPluginDetails;
function getDownloadablePluginDetails(packageJson, downloadUrl, lastUpdated) {
    const details = getPluginDetails(packageJson);
    return {
        ...details,
        isBundled: false,
        isActivatable: false,
        downloadUrl,
        lastUpdated,
    };
}
exports.getDownloadablePluginDetails = getDownloadablePluginDetails;
function getPluginDetailsV1(packageJson) {
    var _a;
    return {
        specVersion: 1,
        name: packageJson.name,
        version: packageJson.version,
        main: 'dist/bundle.js',
        source: packageJson.main,
        id: packageJson.name,
        gatekeeper: packageJson.gatekeeper,
        icon: packageJson.icon,
        title: packageJson.title || packageJson.name,
        description: packageJson.description,
        category: packageJson.category,
        bugs: packageJson.bugs,
        flipperSDKVersion: (_a = packageJson === null || packageJson === void 0 ? void 0 : packageJson.peerDependencies) === null || _a === void 0 ? void 0 : _a['flipper-plugin'],
        pluginType: packageJson === null || packageJson === void 0 ? void 0 : packageJson.pluginType,
        supportedDevices: packageJson === null || packageJson === void 0 ? void 0 : packageJson.supportedDevices,
        engines: packageJson.engines,
    };
}
function getPluginDetailsV2(packageJson) {
    var _a;
    return {
        specVersion: 2,
        name: packageJson.name,
        version: packageJson.version,
        main: packageJson.main,
        source: packageJson.flipperBundlerEntry,
        id: packageJson.id || packageJson.name,
        gatekeeper: packageJson.gatekeeper,
        icon: packageJson.icon,
        title: packageJson.title || packageJson.id || getTitleFromName(packageJson.name),
        description: packageJson.description,
        category: packageJson.category,
        bugs: packageJson.bugs,
        flipperSDKVersion: (_a = packageJson === null || packageJson === void 0 ? void 0 : packageJson.peerDependencies) === null || _a === void 0 ? void 0 : _a['flipper-plugin'],
        pluginType: packageJson === null || packageJson === void 0 ? void 0 : packageJson.pluginType,
        supportedDevices: packageJson === null || packageJson === void 0 ? void 0 : packageJson.supportedDevices,
        engines: packageJson.engines,
    };
}
function getTitleFromName(name) {
    const prefix = 'flipper-plugin-';
    if (name.startsWith(prefix)) {
        return name.substr(prefix.length);
    }
    return name;
}
