// flipper-plugin-lib/lib/pluginPaths.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPluginDirNameFromPackageName = exports.getPluginVersionInstallationDir = exports.getPluginInstallationDir = exports.getPluginSourceFolders = exports.pluginCacheDir = exports.pluginInstallationDir = exports.legacyPluginInstallationDir = void 0;
const path_1 = __importDefault(require("path"));
const os_1 = require("os");
const fs_extra_1 = __importDefault(require("fs-extra"));
const expand_tilde_1 = __importDefault(require("expand-tilde"));
const flipperDataDir = path_1.default.join(os_1.homedir(), '.flipper');
exports.legacyPluginInstallationDir = path_1.default.join(flipperDataDir, 'thirdparty');
exports.pluginInstallationDir = path_1.default.join(flipperDataDir, 'installed-plugins');
exports.pluginCacheDir = path_1.default.join(flipperDataDir, 'plugins');
async function getPluginSourceFolders() {
    const pluginFolders = [];
    if (process.env.FLIPPER_NO_DEFAULT_PLUGINS) {
        console.log('🥫  Skipping default plugins because "--no-default-plugins" flag provided');
        return pluginFolders;
    }
    const flipperConfigPath = path_1.default.join(os_1.homedir(), '.flipper', 'config.json');
    if (await fs_extra_1.default.pathExists(flipperConfigPath)) {
        const config = await fs_extra_1.default.readJson(flipperConfigPath);
        if (config.pluginPaths) {
            pluginFolders.push(...config.pluginPaths);
        }
    }
    pluginFolders.push(path_1.default.resolve(__dirname, '..', '..', 'plugins', 'public'));
    pluginFolders.push(path_1.default.resolve(__dirname, '..', '..', 'plugins', 'fb'));
    return pluginFolders.map(expand_tilde_1.default).filter(fs_extra_1.default.existsSync);
}
exports.getPluginSourceFolders = getPluginSourceFolders;
function getPluginInstallationDir(name) {
    return path_1.default.join(exports.pluginInstallationDir, getPluginDirNameFromPackageName(name));
}
exports.getPluginInstallationDir = getPluginInstallationDir;
function getPluginVersionInstallationDir(name, version) {
    return path_1.default.join(getPluginInstallationDir(name), version);
}
exports.getPluginVersionInstallationDir = getPluginVersionInstallationDir;
function getPluginDirNameFromPackageName(name) {
    return name.replace('/', '__');
}
exports.getPluginDirNameFromPackageName = getPluginDirNameFromPackageName;
